### ISSUE TITLE
Unsupported Get Log Page 0x02

### ISSUE THREAD

**Issue by wawryk** [spdk/spdk#1745]
During nvmf perf tests there is error:
```
ctrlr.c:1964:nvmf_ctrlr_get_log_page: *ERROR*: Unsupported Get Log Page 0x02
```
We have 16 disks (p4600) with latest FW.
Log:
```11:21:49 [target] Done configuring SPDK NVMeOF Target
11:21:49 [initiator1] Trying to discover: 20.0.0.1:4420
11:21:49 [2021-01-14 11:21:49.641591] ctrlr.c:1964:nvmf_ctrlr_get_log_page: *ERROR*: Unsupported Get Log Page 0x02
11:21:49 [initiator1] Trying to discover: 20.0.0.1:4421
11:21:49 [initiator1] Trying to discover: 20.0.0.1:4422
11:21:49 [initiator1] Trying to discover: 20.0.0.1:4423
11:21:49 [initiator1] Trying to discover: 20.0.0.1:4424
11:21:49 [initiator1] Trying to discover: 20.0.0.1:4425
11:21:49 [initiator1] Trying to discover: 20.0.0.1:4426
11:21:49 [initiator1] Trying to discover: 20.0.0.1:4427
11:21:49 [initiator1] Trying to discover: 20.0.0.1:4428
11:21:50 [initiator1] Trying to discover: 20.0.0.1:4429
11:21:50 [initiator1] Trying to discover: 20.0.0.1:4430
11:21:50 [initiator1] Trying to discover: 20.0.0.1:4431
11:21:50 [initiator1] Trying to discover: 20.0.0.1:4432
11:21:50 [initiator1] Trying to discover: 20.0.0.1:4433
11:21:50 [initiator1] Trying to discover: 20.0.0.1:4434
11:21:50 [initiator1] Trying to discover: 20.0.0.1:4435
11:21:50 [initiator1] Trying to discover: 20.0.1.1:4420
11:21:50 [2021-01-14 11:21:50.417102] ctrlr.c:1964:nvmf_ctrlr_get_log_page: *ERROR*: Unsupported Get Log Page 0x02
11:21:50 [initiator1] Trying to discover: 20.0.1.1:4421
11:21:50 [initiator1] Trying to discover: 20.0.1.1:4422
11:21:50 [initiator1] Trying to discover: 20.0.1.1:4423
11:21:50 [initiator1] Trying to discover: 20.0.1.1:4424
11:21:50 [initiator1] Trying to discover: 20.0.1.1:4425
11:21:50 [initiator1] Trying to discover: 20.0.1.1:4426
11:21:50 [initiator1] Trying to discover: 20.0.1.1:4427
11:21:50 [initiator1] Trying to discover: 20.0.1.1:4428
11:21:50 [initiator1] Trying to discover: 20.0.1.1:4429
11:21:50 [initiator1] Trying to discover: 20.0.1.1:4430
11:21:50 [initiator1] Trying to discover: 20.0.1.1:4431
11:21:50 [initiator1] Trying to discover: 20.0.1.1:4432
11:21:50 [initiator1] Trying to discover: 20.0.1.1:4433
11:21:51 [initiator1] Trying to discover: 20.0.1.1:4434
11:21:51 [initiator1] Trying to discover: 20.0.1.1:4435
11:21:51 [initiator1] Found matching subsystems on target side:
11:21:51 [initiator1] ('4420', 'nqn.2018-09.io.spdk:cnode0', '20.0.0.1')
11:21:51 [initiator1] ('4420', 'nqn.2018-09.io.spdk:cnode1', '20.0.0.1')
11:21:51 [initiator1] ('4420', 'nqn.2018-09.io.spdk:cnode2', '20.0.0.1')
11:21:51 [initiator1] ('4420', 'nqn.2018-09.io.spdk:cnode3', '20.0.0.1')
11:21:51 [initiator1] ('4420', 'nqn.2018-09.io.spdk:cnode4', '20.0.1.1')
11:21:51 [initiator1] ('4420', 'nqn.2018-09.io.spdk:cnode5', '20.0.1.1')
11:21:51 [initiator1] ('4420', 'nqn.2018-09.io.spdk:cnode6', '20.0.1.1')
11:21:51 [initiator1] ('4420', 'nqn.2018-09.io.spdk:cnode7', '20.0.1.1')
11:21:51 [initiator1] Below connection attempts may result in error messages, this is expected!
11:21:51 [initiator1] Trying to connect 4420 nqn.2018-09.io.spdk:cnode0 20.0.0.1
11:21:57 [initiator1] Trying to connect 4420 nqn.2018-09.io.spdk:cnode1 20.0.0.1
11:22:04 [initiator1] Trying to connect 4420 nqn.2018-09.io.spdk:cnode2 20.0.0.1
11:22:10 [initiator1] Trying to connect 4420 nqn.2018-09.io.spdk:cnode3 20.0.0.1
11:22:16 [initiator1] Trying to connect 4420 nqn.2018-09.io.spdk:cnode4 20.0.1.1
11:22:23 [initiator1] Trying to connect 4420 nqn.2018-09.io.spdk:cnode5 20.0.1.1
11:22:30 [initiator1] Trying to connect 4420 nqn.2018-09.io.spdk:cnode6 20.0.1.1
11:22:36 [initiator1] Trying to connect 4420 nqn.2018-09.io.spdk:cnode7 20.0.1.1
```
## Current Behavior
It looks like everything works fine (tests pass), but in logs we can see error as above.

## Steps to Reproduce

Commands called on target:
```
sudo ./build/bin/nvmf_tgt &
sudo ./scripts/rpc.py nvmf_create_transport -t tcp -u 8192 -m 4 -c 0
sudo ./scripts/rpc.py bdev_malloc_create -b Malloc0 512 512
sudo ./scripts/rpc.py nvmf_create_subsystem nqn.2016-06.io.spdk:cnode1 -a -s SPDK00000000000001 -d SPDK_Controller1
sudo ./scripts/rpc.py nvmf_subsystem_add_ns nqn.2016-06.io.spdk:cnode1 Malloc0
sudo scripts/rpc.py nvmf_subsystem_add_listener nqn.2016-06.io.spdk:cnode1 -t tcp -a 20.0.0.1 -s 4420
```
Then I call `discover` on initiator:
```
sudo nvme discover -t tcp -s 4420 nqn.2016-06.io.spdk:cnode0 -a 20.0.0.1
```

## Context (Environment including OS version, SPDK version, etc.)
OS: Fedora33
Kernel: 5.9.14-200.fc33.x86_64

**Comment by jimharris** [comment]
Hi @wawryk,

Can you please provide more specifics?  "Run nvmf perf test" can be interpreted many ways.  :-)

For some reason, we are sending a HEALTH_INFO log page request to a discovery controller which is obviously invalid.  I'm just trying to figure out if this is SPDK or Linux kernel initiator that's sending it.

-Jim

**Comment by jimharris** [comment]
I think I figured out that this is getting triggered by the nvme-cli discovery command which we call from the discover_subsystems function in scripts/perf/nvmf/run_nvmf.py.

Can you try just setting up nvmf target with one subsystem, and run nvme discover command manually and confirm you get this same error message?

I tried this with latest master, and could not reproduce this error message on the 0x2 (HEALTH_INFO) log page.  I'm running nvme-cli v1.7.  Could you check nvme-cli version on this system?

**Comment by jimharris** [comment]
I also tried latest nvme-cli tip (1.13.43.g4bde) and was also unable to reproduce this error message.

**Comment by wawryk** [comment]
Hi @jimharris 
thanks for your qiuck reply.
> Can you try just setting up nvmf target with one subsystem, and run nvme discover command manually and confirm you get this same error message?

I will check this tommorow.

> I tried this with latest master, and could not reproduce this error message on the 0x2 (HEALTH_INFO) log page. I'm running nvme-cli v1.7. Could you check nvme-cli version on this system?

Currently on SM3 we have nvme-cli `version 1.11.1`

> Can you please provide more specifics?

Full log [here](https://10.211.11.84:8080/job/NVMf_perf/job/BenchTC4_spdk/36/NUMJOB=4,QD=64,RW=randwrite,label=SM3/console)

**Comment by jimharris** [comment]
I just checked out v1.11.1 in my nvme-cli repo, rebuilt and tried again, but still no reproduction on my end.

Could you see if this happens with both RDMA and TCP?

**Comment by wawryk** [comment]
Sorry for my late answer.
Yes, this happens with both (RDAM and TCP).

**Comment by jimharris** [comment]
> > Can you try just setting up nvmf target with one subsystem, and run nvme discover command manually and confirm you get this same error message?
> 
> I will check this tommorow.
> 

Hi @wawryk,

Did you get a chance to check this?

Thanks,

Jim

**Comment by wawryk** [comment]
Yes, I have the same error in log with one subsystem, also i checked null block device, and there is the same log.

**Comment by jimharris** [comment]
But did you test this manually, without using the run_nvmf.py script?  I want to rule out anything in run_nvmf.py that could be causing this.

**Comment by wawryk** [comment]
I test it using run_nvmf.py and manually by starting nvmf_tgt app on target, and nvme cli discover command on initiator.
Is it ok, or should I do something else ?

**Comment by jimharris** [comment]
Thanks @wawryk.  So when you do it manually by starting nvmf_tgt app on target and just running nvme-cli discover on initiator and it all works, it still fails?

If so, could you update the original report above to describe that?  It will be easier for people to reproduce if they know they can do it with these manual steps, rather than requiring use of the run_nvmf.py script.

**Comment by wawryk** [comment]
@jimharris 
when I start it manually it works fine.

Commands called on target:
```
sudo ./build/bin/nvmf_tgt &
sudo ./scripts/rpc.py nvmf_create_transport -t tcp -u 8192 -m 4 -c 0
sudo ./scripts/rpc.py bdev_malloc_create -b Malloc0 512 512
sudo ./scripts/rpc.py nvmf_create_subsystem nqn.2016-06.io.spdk:cnode1 -a -s SPDK00000000000001 -d SPDK_Controller1
sudo ./scripts/rpc.py nvmf_subsystem_add_ns nqn.2016-06.io.spdk:cnode1 Malloc0
sudo scripts/rpc.py nvmf_subsystem_add_listener nqn.2016-06.io.spdk:cnode1 -t tcp -a 20.0.0.1 -s 4420
```
Then I call `discover` on initiator:
```
sudo nvme discover -t tcp -s 4420 nqn.2016-06.io.spdk:cnode0 -a 20.0.0.1
```
and on target side there is log:
```
[sys_sgci@supermicro3 spdk]$ [2021-01-27 12:12:13.348092] ctrlr.c:2064:nvmf_ctrlr_get_log_page: *ERROR*: Unsupported Get Log Page 0x02
```
Also  pasted this to original report.

**Comment by jimharris** [comment]
Thanks @wawryk.

I was thinking that maybe specifying the NQN was causing nvme-cli to send the log page 0x2 command.  But I couldn't reproduce that on my local system.  I tried both nvme-cli tip. nvme-cli v1.11.1 and nvme-cli v1.7.

Could you try cloning nvme-cli separately on this system, and run your same test using that nvme-cli instead of the packaged one?

**Comment by wawryk** [comment]
@jimharris 
yes, i cloned it today:
```./nvme --version
nvme version 1.13.56.g102a
```
and on the newest version target show:
```
[2021-01-28 13:27:26.695056] ctrlr.c:2064:nvmf_ctrlr_get_log_page: *ERROR*: Unsupported Get Log Page 0x02
```

**Comment by changpe1** [comment]
@wawryk I could not reproduce this issue in my environment either.

I tested nvme-1.10.1 and nvme-1.14.1.ged95, both could not reproduce this issue.

can you try start SPDK with  `build/bin/nvmf_tgt -L nvmf`, and attach the log ?

**Comment by wawryk** [comment]
@changpe1 on my local setup with one nvme disk, I can not reproduce it too, this case only occurs on setup with 16 nvme drives (which currently have scheduled tests) and this log still occurs there

**Comment by jimharris** [comment]
@wawryk - earlier in this thread, you were able to reproduce this with null backend (no SSDs).  Is that still the case?

**Comment by wawryk** [comment]
with null backend i check it today, and still can be reproduces.
log for @changpe1:
```
sudo ./build/bin/nvmf_tgt -L nvmf &
[1] 3998929
[sys_sgci@supermicro3 spdk]$ [2021-06-02 12:03:51.772984] Starting SPDK v21.04 git sha1 801671015 / DPDK 20.11.0 initialization...
[2021-06-02 12:03:51.773040] [ DPDK EAL parameters: [2021-06-02 12:03:51.773050] nvmf [2021-06-02 12:03:51.773056] --no-shconf [2021-06-02 12:03:51.773065] -c 0x1 [2021-06-02 12:03:51.773070] --log-level=lib.eal:6 [2021-06-02 12:03:51.773078] --log-level=lib.cryptodev:5 [2021-06-02 12:03:51.773084] --log-level=user1:6 [2021-06-02 12:03:51.773091] --base-virtaddr=0x200000000000 [2021-06-02 12:03:51.773099] --match-allocations [2021-06-02 12:03:51.773106] --file-prefix=spdk_pid3998930 [2021-06-02 12:03:51.773113] ]
EAL: No available hugepages reported in hugepages-1048576kB
EAL: No legacy callbacks, legacy socket not created
[2021-06-02 12:03:51.824849] app.c: 535:spdk_app_start: *NOTICE*: Total cores available: 1
[2021-06-02 12:03:51.968559] reactor.c: 929:reactor_run: *NOTICE*: Reactor started on core 0
[2021-06-02 12:03:51.968622] accel_engine.c: 852:spdk_accel_engine_initialize: *NOTICE*: Accel engine initialized to use software engine.
```

Then i do the same thing as previous: 
```
sudo ./scripts/rpc.py nvmf_create_transport -t tcp -u 8192 -m 4 -c 0
sudo ./scripts/rpc.py bdev_malloc_create -b Malloc0 512 512
sudo ./scripts/rpc.py nvmf_create_subsystem nqn.2016-06.io.spdk:cnode1 -a -s SPDK00000000000001 -d SPDK_Controller1
sudo ./scripts/rpc.py nvmf_subsystem_add_ns nqn.2016-06.io.spdk:cnode1 Malloc0
sudo scripts/rpc.py nvmf_subsystem_add_listener nqn.2016-06.io.spdk:cnode1 -t tcp -a 20.0.0.1 -s 4420
```
When use nvme discover command on initiator, then i see log:
```
[sys_sgci@supermicro3 spdk]$ [2021-06-02 12:08:37.967731] ctrlr.c:2075:nvmf_ctrlr_get_log_page: *ERROR*: Unsupported Get Log Page 0x02
```

nvme version on target and initiator:
```
nvme version 1.13.56.g102a
```

**Comment by jimharris** [comment]
@ksztyber or @tomzawadzki, could one of you try getting on this system with @wawryk, and walk through a debug version of fio in gdb for this nvme discover command?

I think you'll want to run nvmf_tgt in gdb too, set a breakpoint on the ctrlr.c:2075 error message, then when it breaks, break into nvme-cli, do a backtrace and see how nvme-cli is sending a SMART log page request as part of an nvme-discover.

Looking at the code, I see no way this could happen - it's going to need active debugging on this one system that exhibits the issue.

@wawryk - I noticed you name the subsystem 'cnode0' but then try to discover 'cnode1'.  You shouldn't even need to specify a subnqn at all.  What happens if you omit the subnqn completely?

**Comment by ksztyber** [comment]
Sure @jimharris, I can take a look at that.

**Comment by ksztyber** [comment]
After taking a closer look, it seems that it can be reproduced without nvme-cli:
```
echo 'nqn=nqn.2014-08.org.nvmexpress.discovery,transport=tcp,traddr=20.0.0.1,trsvcid=4420' > /dev/nvme-fabrics
```
And it appears to be a kernel bug that has already been fixed: https://lore.kernel.org/linux-nvme/20210114000022.1641939-1-sagi@grimberg.me/

I think it got merged in v5.11 and got backported to v5.10, so this shouldn't be happening on any of the newer kernels.
